service/worker/batcher/activities.go

```diff
@@ -41,6 +42,258 @@ var (
 	errNamespaceMismatch = errors.New("namespace mismatch")
 )
 
+// batchProcessorConfig holds the configuration for batch processing
+type batchProcessorConfig struct {
+	namespace         string
+	adjustedQuery     string
+	rps               float64
+	concurrency       int
+	initialPageToken  []byte
+	initialExecutions []*commonpb.WorkflowExecution
+}
+
+// batchWorkerProcessor defines the interface for different worker processor types
+type batchWorkerProcessor func(
+	ctx context.Context,
+	taskCh chan taskDetail,
+	respCh chan taskResponse,
+	rateLimiter *rate.Limiter,
+	sdkClient sdkclient.Client,
+	frontendClient workflowservice.WorkflowServiceClient,
+	metricsHandler metrics.Handler,
+	logger log.Logger,
+)
+
+// Task represents a task to be processed (alias for taskDetail for consistency with pseudocode)
+type Task = taskDetail
+
+// Page represents a page of workflow executions to be processed
+type Page struct {
+	executions     []*commonpb.WorkflowExecution
+	submittedCount int
+	successCount   int
+	errorCount     int
+	nextPageToken  []byte
+	pageNumber     int
+	prev, next     *Page
+}
+
+// hasNext returns true if there are more pages to fetch
+func (p *Page) hasNext() bool {
+	return len(p.nextPageToken) > 0
+}
+
+// allSubmitted returns true if all executions in this page have been submitted
+func (p *Page) allSubmitted() bool {
+	return p.submittedCount == len(p.executions)
+}
+
+// nextTask returns the next task to be submitted from this page
+func (p *Page) nextTask(hbd HeartBeatDetails) Task {
+	if p.submittedCount >= len(p.executions) {
+		return Task{} // No more tasks in this page
+	}
+
+	task := Task{
+		execution:  p.executions[p.submittedCount],
+		attempts:   1,
+		hbd:        hbd,
```

Comment:
I think you don't need `hbd` on the task here or `pageNumber`, you can pass a reference to the `page` instead.